File: Ilya-Pankratov/Fourth term/Block 3/Exam/ExamLib/HashTables/StripedCuckooHashSet.cs
namespace ExamLib;

public class StripedCuckooHashSet<T> : PhasedCuckooHashSet<T>
{
    Mutex[,] locks;

    public StripedCuckooHashSet(int capacity) : base(capacity)
    {
        locks = new Mutex[2, capacity];
        for (var i = 0; i < 2; i++)
        {
            for (var j = 0; j < capacity; j++)
            {
                locks[i, j] = new Mutex();
            }
        }
    }

    public StripedCuckooHashSet(int capacity, IEqualityComparer<T> comparer) : this(capacity)
    {
        this.comparer = comparer;
    }

    protected override void Acquire(T x)
    {
        locks[0, Hash0(x) % locks.GetLength(1)].WaitOne();
        locks[1, Hash1(x) % locks.GetLength(1)].WaitOne();
    }

    protected override void Release(T x)
    {
        locks[0, Hash0(x) % locks.GetLength(1)].ReleaseMutex();
        locks[1, Hash1(x) % locks.GetLength(1)].ReleaseMutex();
    }

    protected override void Resize()
    {
        var oldCapacity = capacity;
        for (var i = 0; i < locks.GetLength(1); i++)
        {
            locks[0, i].WaitOne();
        }

        try
        {
            if (capacity != oldCapacity)
            {
                return;
            }

            var oldTable = table;
            capacity = 2 * capacity;
            table = new List<T>[2, capacity];

            for (var i = 0; i < 2; i++)
            {
                for (var j = 0; j < capacity; j++)
                {
                    table[i, j] = new List<T>(ListSize);
                }
            }

            for (var i = 0; i < 2; i++)
            {
                for (var j = 0; j < oldCapacity; j++)
                {
                    foreach (var z in oldTable[i, j])
                    {
                        Add(z);
                    }
                }
            }
        }
        finally
        {
            for (var i = 0; i < locks.GetLength(1); i++)
            {
                locks[0, i].ReleaseMutex();
            }
        }
    }
}

File: Khelmianov-Artem/Term4/Block3/Dekanat/src/main/kotlin/collections/StripedCuckooHashSet.kt
package collections

import java.util.concurrent.locks.ReentrantLock

class StripedCuckooHashSet<T : Any>(capacity: Int = 64) : PhasedCuckooHashSet<T>(capacity) {

    private val locks = Array(2) { Array(capacity) { ReentrantLock() } }

    override fun acquire(element: T) {
        locks[0, element.hash(0, modulus = locks.size)].lock()
        locks[1, element.hash(1, modulus = locks.size)].lock()
    }

    override fun release(element: T) {
        locks[0, element.hash(0, modulus = locks.size)].unlock()
        locks[1, element.hash(1, modulus = locks.size)].unlock()
    }

    override fun resize() {
        val oldCapacity = capacity
        locks[0].forEach { it.lock() }
        try {
            if (capacity != oldCapacity) return
            val oldStorage = storage
            capacity *= 2
            storage = Array(2) { Array(capacity) { mutableListOf<T>() } }
            oldStorage.forEach { list -> list.forEach { this.add(it) } }
        } finally {
            locks[0].forEach { it.unlock() }
        }
    }
}